rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp

```diff
@@ -88,6 +89,15 @@ class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
     if (RCL_RET_OK != ret) {
       throw std::runtime_error("SubscriptionIntraProcess init error initializing guard condition");
     }
+
+    TRACEPOINT(
+      rclcpp_subscription_callback_added,
+      (const void *)subscription_handle,
```

Comment:
Why do you need the handle?

Reply:
I don't have a lot of context, but maybe there's a better thing to put here like the intra process subscription's address? I'm not sure what's being traced in this case.

Reply:
@wjwwood this is to link the callback object to the subscription (and on). The subscription itself is linked to its node when initializing the subscription in `rcl` here: https://github.com/ros2/rcl/blob/9034c406c2d5bc4ec291f8e1ad4303cde0184cbc/rcl/src/rcl/subscription.c#L188

This is how we can "link" tracepoint events together to get a big picture when performing analyses. See a simple representation of the output data [here](https://gitlab.com/micro-ROS/ros_tracing/ros2_tracing/snippets/1912239#L57). Events are in separate dataframes, but we can link a callback instance back to its node, e.g.

* callback instance at line 143
* linked back to its callback object at line 129 through its callback_object (which is the address of the `AnySubscriptionCallback` object)
* linked back to its subscription at line 83 through its subscription handle (`rcl_subscription_t` pointer)
* linked back to its node at line 67 through its node handle (`rcl_node_t` pointer)

We can't know _all_ of this information just by putting a tracepoint here in `SubscriptionIntraProcess`. Let me know if this makes sense.

Reply:
But would it not make sense to "attach" the rcl subscription handle to the intra process subscription object, and then the any subscription callback to the intra process subscription object?

It feels wrong to me to change the interface of the intra process subscription object just for tracing...

Reply:
> But would it not make sense to "attach" the rcl subscription handle to the intra process subscription object, and then the any subscription callback to the intra process subscription object?

That does make sense; I think that would work, yes.

So we'd have something like this for both `Subscription` and `SubscriptionIntraProcess`:

* `rcl_subscription_t` handle <=> subscription object
* subscription object <=> `AnySubscriptionCallback` object

And the first step could be done in the `Subscription` constructor for both subscription classes.

> It feels wrong to me to change the interface of the intra process subscription object just for tracing...

Of course, I agree. I was just trying to find a solution :smiley:

Reply:
@wjwwood the only thing is that it requires a new tracepoint, i.e. a modification to `tracetools` and thus an update to the `ros2_tracing` package for this bug fix.

Reply:
Unfortunately, I think if `ros2_tracing` is so tightly coupled to the internal layout of rclcpp then we have no choice but to update them together... Maybe there's a better way to decouple them, but I still don't think that's a reason to mutate the rclcpp API if it can be avoided.

Reply:
@wjwwood for sure. I've made the changes with 03c0a89bd5c674f0909536647a96cf89b27e6307.

This should still pass CI, even if `tracetools`/`ros2_tracing` has not been updated (because the tracepoint calls simply get preprocessed out here, and the tracing tests are only run on the CI over on GitLab). What is the deadline for updating the version of the `ros2_tracing` package used in ros2/ros2 for Eloquent? I'm working on it, but it might not be ready until this weekend.

Reply:
The hard deadline is Tuesday I believe, but you should aim to get all this in by Monday just in case there are issues.